File: typescript/src/client/token-store.ts
/**
 * Pluggable token persistence.
 *
 * The client stores three pieces of state:
 *
 *   - `token`: the encoded LIC1 string returned by `/activate` or `/refresh`.
 *   - `graceStartSec`: absolute unix-seconds timestamp when the client
 *     entered unreachable-grace, or null when not in grace.
 *
 * Storage is a single JSON blob so reads are atomic (one file read, one
 * parse) and a partial write can never leave us with a token without its
 * matching grace state.
 *
 * `FileTokenStore` is the default for Node/Bun: writes go via `rename` of
 * a temp file so power loss can't corrupt the store. A
 * `MemoryTokenStore` is provided for tests and for consumers who want to
 * inject their own persistence (e.g. OS keychain) via the `TokenStore`
 * interface.
 */

import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';

export interface StoredTokenState {
  readonly token: string | null;
  /** Unix seconds when grace started, null when not in grace. */
  readonly graceStartSec: number | null;
}

export const EMPTY_STATE: StoredTokenState = {
  token: null,
  graceStartSec: null,
};

export interface TokenStore {
  /** Read the full state. Returns {@link EMPTY_STATE} when no token has been
   *  stored yet — implementations SHOULD NOT throw on "file not found". */
  read(): Promise<StoredTokenState>;
  /** Replace the entire state atomically. */
  write(state: StoredTokenState): Promise<void>;
  /** Clear all state. Equivalent to `write(EMPTY_STATE)` for most stores;
   *  kept separate so implementations can drop keychain entries, etc. */
  clear(): Promise<void>;
}

// ---------- FileTokenStore ----------

/**
 * Default file-backed token store. Reads a JSON document from `path`,
 * writes via tmp-file + rename for atomicity, and creates the parent
 * directory on first write.
 *
 * File layout:
 *
 *   {
 *     "token": "LIC1.<header>.<payload>.<sig>" | null,
 *     "graceStartSec": 1746000000 | null
 *   }
 *
 * Unknown top-level fields are preserved on write so a future client
 * version that adds fields doesn't wipe them — but we only READ the ones
 * we know about. This keeps forward-compat cheap without forcing schema
 * migrations for every added field.
 */
export class FileTokenStore implements TokenStore {
  readonly #path: string;

  constructor(path: string) {
    if (path.length === 0) throw new Error('FileTokenStore path must be non-empty');
    this.#path = path;
  }

  async read(): Promise<StoredTokenState> {
    try {
      const raw = await readFile(this.#path, 'utf8');
      const parsed = JSON.parse(raw) as Record<string, unknown>;
      return {
        token: typeof parsed.token === 'string' ? parsed.token : null,
        graceStartSec:
          typeof parsed.graceStartSec === 'number' && Number.isFinite(parsed.graceStartSec)
            ? parsed.graceStartSec
            : null,
      };
    } catch (err) {
      if (isNotFound(err)) return EMPTY_STATE;
      // Corrupted JSON or unreadable file — surface as "no token" rather than
      // throwing, so a bricked store doesn't brick the whole client. The
      // caller can observe the corruption via a subsequent `validate()`
      // that returns `NoToken`.
      return EMPTY_STATE;
    }
  }

  async write(state: StoredTokenState): Promise<void> {
    const dir = dirname(this.#path);
    await mkdir(dir, { recursive: true });
    const tmp = `${this.#path}.tmp-${process.pid}-${Date.now().toString(36)}`;
    const body = `${JSON.stringify(
      { token: state.token, graceStartSec: state.graceStartSec },
      null,
      2,
    )}\n`;
    await writeFile(tmp, body, { mode: 0o600 });
    await rename(tmp, this.#path);
  }

  async clear(): Promise<void> {
    try {
      await unlink(this.#path);
    } catch (err) {
      if (!isNotFound(err)) throw err;
    }
  }
}

// ---------- MemoryTokenStore ----------

/**
 * In-memory token store. Suitable for tests and for ephemeral runtimes
 * (CLIs that re-activate on every invocation). Thread-safe only to the
 * extent that JS itself is single-threaded per event loop.
 */
export class MemoryTokenStore implements TokenStore {
  #state: StoredTokenState = EMPTY_STATE;

  async read(): Promise<StoredTokenState> {
    return this.#state;
  }

  async write(state: StoredTokenState): Promise<void> {
    this.#state = state;
  }

  async clear(): Promise<void> {
    this.#state = EMPTY_STATE;
  }
}

// ---------- internals ----------

function isNotFound(err: unknown): boolean {
  return typeof err === 'object' && err !== null && (err as { code?: string }).code === 'ENOENT';
}

File: licensing/client/token_store.go
package client

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

// StoredTokenState is the shape persisted by TokenStore implementations.
type StoredTokenState struct {
	GraceStartSec *int64 `json:"graceStartSec"`
	Token         string `json:"token"`
}

// EmptyState returns a zero-value state (no token, no grace).
func EmptyState() StoredTokenState {
	return StoredTokenState{}
}

// TokenStore is the pluggable persistence interface for the client's
// current token and grace-start timestamp.
type TokenStore interface {
	Read() (StoredTokenState, error)
	Write(state StoredTokenState) error
	Clear() error
}

// ---------- MemoryTokenStore ----------

// MemoryTokenStore is an in-memory TokenStore suitable for tests.
type MemoryTokenStore struct {
	state StoredTokenState
	mu    sync.Mutex
}

// NewMemoryTokenStore constructs a new memory token store.
func NewMemoryTokenStore() *MemoryTokenStore {
	return &MemoryTokenStore{}
}

func (m *MemoryTokenStore) Read() (StoredTokenState, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.state, nil
}

func (m *MemoryTokenStore) Write(state StoredTokenState) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.state = state
	return nil
}

// Clear resets the in-memory state to empty. Safe for concurrent use.
func (m *MemoryTokenStore) Clear() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.state = EmptyState()
	return nil
}

// ---------- FileTokenStore ----------

// FileTokenStore persists token state to a JSON file. Atomic writes via
// tmp+rename. Creates parent directories on write. Mode 0600.
type FileTokenStore struct {
	path string
}

// NewFileTokenStore constructs a new file token store.
func NewFileTokenStore(path string) *FileTokenStore {
	return &FileTokenStore{path: path}
}

func (f *FileTokenStore) Read() (StoredTokenState, error) {
	data, err := os.ReadFile(f.path)
	if err != nil {
		if os.IsNotExist(err) {
			return EmptyState(), nil
		}
		return EmptyState(), err
	}
	var state StoredTokenState
	if err := json.Unmarshal(data, &state); err != nil {
		// Surface the error. Silently degrading to empty state hides
		// tampering and forces a re-activate (seat spend). Callers decide
		// whether to Clear() and re-activate or surface to the user.
		return EmptyState(), fmt.Errorf("token store file is corrupt: %w", err)
	}
	return state, nil
}

func (f *FileTokenStore) Write(state StoredTokenState) error {
	data, err := json.Marshal(state)
	if err != nil {
		return err
	}
	dir := filepath.Dir(f.path)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return err
	}
	// Atomic + durable: write to tmp, fsync file, rename, fsync dir.
	tmp := f.path + ".tmp"
	tf, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
	if err != nil {
		return err
	}
	if _, err := tf.Write(data); err != nil {
		_ = tf.Close()
		_ = os.Remove(tmp)
		return err
	}
	if err := tf.Sync(); err != nil {
		_ = tf.Close()
		_ = os.Remove(tmp)
		return err
	}
	if err := tf.Close(); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	if err := os.Rename(tmp, f.path); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	// Best-effort fsync of the directory to persist the rename.
	if d, err := os.Open(dir); err == nil {
		_ = d.Sync()
		_ = d.Close()
	}
	return nil
}

// Clear deletes the backing file; a missing file is treated as success.
func (f *FileTokenStore) Clear() error {
	err := os.Remove(f.path)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}
